Improve experience of working with node builtins

### Type of issue

Feature request

### Description

Using node builtins currently requires employing some hacks due to the way we manage tree shaking in conjunction with our webpack node configuration. We should improve the experience of using node builtins.

### Current behavior

```js
// some-plugin.js
import fs from 'fs';
import {createPlugin} from 'fusion-core';

export default createPlugin({
  deps: {},
  provides: () => {
    // do some things with fs...
  }
}}

// src/main.js
import somePlugin from './some-plugin';
__NODE__ && app.register(somePlugin):
```

Running `dev` results in: 

```
error: (client) ./src/some-plugin.js
Module not found: Error: Can't resolve 'fs' in ...
```

This happens for two reasons.
1. we don't tree shake in development, so `src/some-plugin.js` is included in the client bundle
2. we set the `node.fs` webpack config to be `false`

### Expected behavior

The build should work, but the plugin won't be registered.

### Steps to reproduce

1. scaffold a fusion app
2. create a plugin that imports fs
3. apply that plugin in __NODE__ only
4. run `fusion dev`

### Proposed Solution

Long term, we can investigate ways of supporting tree shaking in development. Short term, we should set all `node` builtins to be `empty` (where possible) for the client build. This will allow the user to import them, but with throw errors if the user tries to use them. 

NOTE: Unfortunately, problems often arise when certain node builtins are set to 'empty', specifically surrounding testing, so we may need to make exceptions in certain cases.

--- a/build/compiler.js
+++ b/build/compiler.js
@@ -218,7 +218,7 @@ function getConfig({target, env, dir, watch, cover}) {
         /**
          * Tape requires `fs` to be defined
          */
-        fs: env === 'test' && target === 'web' ? 'empty' : false,
+        fs: target === 'web' ? 'empty' : false,
       },
       node
     ),

--- a/test/cli/dev.js
+++ b/test/cli/dev.js
@@ -151,6 +151,22 @@ test('`fusion dev` server render error', async t => {
   t.end();
 });
 
+test('`fusion dev` works with fs', async t => {
+  const dir = path.resolve(__dirname, '../fixtures/tree-shaking');
+  const {proc, port} = await dev(`--dir=${dir}`);
+  const res = await request(`http://localhost:${port}/fs`);
+  t.ok(res.includes('writeFile'), 'supports fs api on the server');
+  const mainRes = await request(
+    `http://localhost:${port}/_static/client-main.js`
+  );
+  t.ok(
+    mainRes.includes('node-libs-browser/mock/empty.js'),
+    'includes empty fs for browser in dev'
+  );
+  proc.kill();
+  t.end();
+});
+
 test('`fusion dev` recovering from errors', async t => {
   const dir = path.resolve(__dirname, '../fixtures/server-startup-error');
   const {res, proc} = await dev(`--dir=${dir}`, {

--- a/test/fixtures/tree-shaking/src/main.js
+++ b/test/fixtures/tree-shaking/src/main.js
@@ -1,0 +1,18 @@
+import App from 'fusion-core';
+import fs from 'fs';
+
+export default async function() {
+  const app = new App('element', el => el);
+  __NODE__ &&
+    app.middleware((ctx, next) => {
+      if (ctx.url === '/fs') {
+        ctx.body = Object.keys(fs);
+      }
+      return next();
+    });
+  __BROWSER__ &&
+    app.middleware((ctx, next) => {
+      console.log(Object.keys(fs));
+    });
+  return app;
+}

